### ISSUE TITLE
Using SHA checksum for testing results is insustainable

### ISSUE THREAD

**Issue by llaniewski** [CFD-GO/TCLB#213]
As we go into testing on multiple platforms (cuda/cpu, different compilers etc), the comparison based on comparing checksums of `vti` files cannot work. The result files have to be compared as numerical fields, up to a certain precision.

Currently running the `tools/tests.sh` fails on most architectures, as the sha sums are from Travis-CI VMs.

Possible solutions:
- base comparison on absolute difference between fields below some epsilon, done with:
   - `vtk` modules in `python` (lot of potential compatibility issues there)
   - some quick write up of a simple `vti` loader for `R`/`python` (not based of vtk, and no support for other formats)
   - `pvpython`?
- base comparison only on other types of output (TXT, Log, etc)

**Comment by ggruszczynski** [comment]
We have faced the same problem recently.
I think that the `tools/tests.sh` shall call a python script which would compare output and expected .vti files, ie. Solution 1a). Compatibility issues occur with g++, nvcc, etc as well ;-)
They can be mitigated installing a **predefined** environment, like
```
$ apt-get install python3.6
$ pip3 install somepackage==1.2.3
```
Having the data read into an array, we can _assert_allclose_ similar to:
https://github.com/CFD-GO/TCLB_tools/blob/master/Python/symbolic_tools/tests/test_vti_reader.py

**Comment by mdzik** [comment]
I would say that test's that run on build machine are too short (to few iterations) to use "epsilon" - we could miss something. I could rewrite the test to include the platform(whole toolchain in fact) in test reference. So CUDA will be tested against CUDA.sha etc. And the author needs to provide the reference "by hand"/"by script".

My idea was that you test manually if solver work (convergence, reference solution etc.) and then you generate ci-test that covers ~max 1 minutes of computations and store the result. If anything changes in output manual reevaluation is needed. The problem is that repo structure does not handle it well and we lack scripts to handle that.

And there is still option to not include VTI just leave CSV as tests.

My proposition would be to create ./test.sh [model] [platform] [mode] script were:

- platform if predefined based on Travis input?
- mode is short/long where it changes number of iterations used in .xml (so, in fact, xml.Rt or something similarly crazy ;) )

I think that I would also need to provide script to run test env (travis) locally which is perfectly doable

And I would vote to include "tests" in TCLB main repo - most repos have it like that.

**Comment by mdzik** [comment]
For those interested, instruction below. Let me know if we want to use such magic - it could be easier and I could make travis-derived docker images and scripts to create tests reference values


I assume that:

- you have working docker (install after google, check if 'docker ps' doesn't fail)

And magic comes:

```bash
$ docker pull travisci/ci-sardonyx:packer-1553530531-f909ac5
$ docker ls 
REPOSITORY             TAG                         IMAGE ID            CREATED             SIZE
travisci/ci-sardonyx   packer-1553530531-f909ac5   SOMETHINHERE        5 months ago        9.07GB
```
if you use proxy:
```
$docker run -it  --env http_proxy=$http_proxy --env https_proxy=$https_proxy --network host SOMETHINHERE /bin/bash
```
else:
```
$docker run -it  SOMETHINHERE /bin/bash
```

then, inside docker container (you should be inside after docker run)
```
aptitude update
apt install ruby-dev  ruby
git clone https://github.com/travis-ci/travis-build
cd travis-build
mkdir -p ~/.travis
ln -s $PWD ~/.travis/travis-build
gem install bundler
bundle install --gemfile ~/.travis/travis-build/Gemfile
bundler binstubs travis
```
Then, create script:
```bash
#/bin/bash
REPO=mdzik/TCLB.git
BRANCH="d3q27_csf"
BUILD="180.10" #check this on travis site, only second number is important (model)
set -e
set -x
cd /
rm -rf ~/build
rm -rf TCLB_build
mkdir TCLB_build
cd TCLB_build
git clone --branch=$BRANCH  https://github.com/$REPO
cd TCLB
##cd $(echo $REPO | sed 's/\.git//g')
~/.travis/travis-build/bin/travis compile --skip-version-check --skip-completion-check -r $REPO --no-interactive $BUILD > run.sh
cd /TCLB_build
cat TCLB/run.sh | sed "s/--branch\\\\=\\\\'\\\\/--branch\\\\=\\\\'$BRANCH\\\\/g" > TCLB/run2.sh
bash TCLB/run2.sh
```

and run it, preferably from /
you could now change sth, push and rerun to see the result. It is possible to change locally - remove rm -rf ~/build line and edit the code there. Then git push or copy it. Some interlay with gitauth (git auth token) might be needed

**Comment by mdzik** [comment]
#269

**Comment by llaniewski** [comment]
Closed by #259
